src/Tools/BuildValidator/CompilationDiff.cs

```diff
@@ -165,9 +176,29 @@ public string MiscErrorMessage
                         assemblyInfo,
                         RebuildResult.BinaryDifference,
                         originalPortableExecutableBytes: originalBytes,
+                        originalPdbReader: compilationFactory.OptionsReader.PdbReader,
                         rebuildPortableExecutableBytes: rebuildBytes,
+                        rebuildPdbReader: getRebuildPdbReader(),
                         rebuildCompilation: rebuildCompilation);
                 }
+
+                unsafe MetadataReader getRebuildPdbReader()
+                {
+                    if (hasEmbeddedPdb)
+                    {
+                        fixed (byte* rebuildPtr = rebuildBytes)
+                        {
+                            var peReader = new PEReader(rebuildPtr, rebuildBytes.Length);
+                            return peReader.GetEmbeddedPdbMetadataReader() ?? throw ExceptionUtilities.Unreachable;
```

Comment:
As you noted in our chat I believe this is the root of the problem. The pointer `rebuildPtr` is not fixed once this `return` executes so this is a memory corruption issue waiting to happen. 

I think it makes the most sense to take the same approach with portable PDBs. The `CompilationDiff` type should just have a field of type `byte[]?` for the portable PDB fields that will be `null` in the case that there is an embedded PDB. Inside the diff function then we can use `fixed` is is done to day in a safe limited scope for the readers.

My instinct is to encapsulate this data into a  `record` that `CompilationDiff` can use: 

```c#
record AssemblyRawInfo(byte[] PEBytes, byte[]? PortablePdbBytes) 
{
   public bool HasEmbeddedPdb => PortablePdbBytes is null;
}
```

Reply:
I have eliminated the use of 'fixed' here although it didn't seem directly connected to any bad behavior. Just using the Stream and ImmutableArray overloads to create PEReader/MetadataReader seems adequate.